storage: quiesce ranges

petermattis opened the pull request:
Cc @cockroachdb/stability, @arjunravinarayan 

[This started as small experiment to quantify what percentage of ranges could be quiesced. In `block_writer` and `zerosum` testing the answer is approximately "all of them". The heartbeat numbers appear too good, dropping to double digits even when there are thousands of ranges. That makes me suspect I'm doing something wrong, but I haven't been able to figure out what yet. Regardless, consider this a starting point for discussion about using quiescence as an alternative or in addition to coalescing heartbeats. Surprisingly, all of the tests pass.]

Quiesce ranges when there is no pending work. A quiesced range is not
ticked and thus doesn't create MsgHeartbeat requests or cause
elections. The Raft leader for a range initiates quiescence just before
ticking the local replica and checks various conditions: no pending raft
commands, no pending raft ready, all of the followers are up to date,
etc. Quiescence is initiated by a special MsgHeartbeat that is tagged as
Quiesce. Upon receipt, the follower checks to see if the term/commit
matches and marks the local replica as quiescent. If the term/commit do
not match the MsgHeartbeat is passed through to Raft which will generate
a MsgHeartbeatResp that will unquiesce the sender.

Any Raft operation on the local replica will unquiesce the Replica. For
example, a Raft operation initiated on a follower will unquiesce the
follower which will send a MsgProp to the leader that will unquiesce the
leader. If the leader of a quiesced range dies, followers will not
notice, though any request directed to the range will eventually end up
on a follower which will unquiesce the follower and lead to an election.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9315)

<!-- Reviewable:end -->

petermattis committed:
storage: quiesce ranges

Quiesce ranges when there is no pending work. A quiesced range is not
ticked and thus doesn't create MsgHeartbeat requests or cause
elections. The Raft leader for a range initiates quiescence just before
ticking the local replica and checks various conditions: no pending raft
commands, no pending raft ready, all of the followers are up to date,
etc. Quiescence is initiated by a special MsgHeartbeat that is tagged as
Quiesce. Upon receipt, the follower checks to see if the term/commit
matches and marks the local replica as quiescent. If the term/commit do
not match the MsgHeartbeat is passed through to Raft which will generate
a MsgHeartbeatResp that will unquiesce the sender.

Any Raft operation on the local replica will unquiesce the Replica. For
example, a Raft operation initiated on a follower will unquiesce the
follower which will send a MsgProp to the leader that will unquiesce the
leader. If the leader of a quiesced range dies, followers will not
notice, though any request directed to the range will eventually end up
on a follower which will unquiesce the follower and lead to an election.

bdarnell commented:
Nice! This is simpler and much more effective than expected (block_writer, with its uniform random distribution, is something of a worst case for quiescing ranges).

Interestingly, gamma currently shows that a substantial fraction of the replicas on a node have active leases, so we're able to quiesce replicas faster than the lease expiration by checking on every tick. This kind of makes sense - on a fast network, unless your traffic is so heavy that the replica ~always has a pending command, you're unlikely to have anything to do when the raft tick interval is reached. The benefits of this change will be reduced with higher network latency, perhaps by a lot.

This looks like a pretty good change to me overall. It would be interesting to deploy this to gamma and see how the reduction in heartbeats impacts the numbers we actually care about (both throughput and latency, especially latency in the presence of failures once we have chaos set up on the test clusters)

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1998 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRXg1zjtxy0RNGHO0V2:-KRXg1zjtxy0RNGHO0V3:bp4gma9) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcacfeff65134dd2f0e274871f069683083239e0/storage/replica.go#L1998)):*
> ```Go
> 		return false
> 	}
> 	for id, progress := range status.Progress {
> ```

if `r.mu.replicaID` is not in `status.Progress`, we shouldn't quiesce. (I think that's possible in weird edge cases during replica changes)

---

*[storage/replica.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRXeVB-Hp_lPN31eAL-:-KRXeVB-Hp_lPN31eAL0:b-n3mygq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcacfeff65134dd2f0e274871f069683083239e0/storage/replica.go#L2001)):*
> ```Go
> 		if progress.Match != status.Applied {
> 			if log.V(3) {
> 				log.Infof(r.ctx, "not quiescing: replica %d behind (%d) != applied (%d)",
> ```

The replica could also be *ahead*: when a new command is proposed, the leader's Progress.Match is advanced to the LastIndex, so this is how we would tell that there are uncommitted entries in the log (especially the dummy entry that concludes an election; others should appear in `pendingCmds`). It's probably worth a comment that we want all of Applied, Commit, LastIndex, and all Progress.Match indexes to be equal, but we get the LastIndex check implicitly via Match (or we could add an explicit check for LastIndex == Commit).

---

*[storage/store.go, line 2294 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRXhfjKDPp77C3L62ek:-KRXhfjLAf7aUEd_a9XZ:b-q0lk1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcacfeff65134dd2f0e274871f069683083239e0/storage/store.go#L2294)):*
> ```Go
> 		}
> 		status := r.RaftStatus()
> 		// TODO(peter): Is checking status.{Term,Commit} correct here?
> ```

Yes, I think so. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-:-KRXjfjfhCMO9NiNjbNV:bj6dld8)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Ok, I think I've figured out the unexpectedly low heartbeat numbers for `zerosum`. I had forgotten that tool uses a zipf distribution for selecting which accounts to perform transfers between. So ranges are not accessed uniformly.

For `block_writer`, I expect the number of ranges accessed per second to be approximately the number of requests per second. The usual case for quiescence is that we'll quiesce a newly active range before it sends a single heartbeat. So if a cluster is performing 5000 reqs/second we should see ~5000 "quiesce" heartbeats/sec and only a handful of heartbeat responses. Compared to pre-quiescence on gamma where we're sending 32k heartbeat requests and responses per second (64k total messages). So 64k messages -> 5k messages. That's my expectation for what benefit we'll see.

bdarnell commented:
We'll send 59k fewer messages per second, but the question is what does that translate to in terms of increased application QPS?

spencerkimball commented:
Here are the numbers for a uniform distribution:

```
    1000 requests/second with     1000 ranges averages 631.00 ranges
    2500 requests/second with     1000 ranges averages 917.80 ranges
    5000 requests/second with     1000 ranges averages 993.90 ranges
   10000 requests/second with     1000 ranges averages 1000.00 ranges
   25000 requests/second with     1000 ranges averages 1000.00 ranges
   50000 requests/second with     1000 ranges averages 1000.00 ranges
    1000 requests/second with     2500 ranges averages 823.30 ranges
    2500 requests/second with     2500 ranges averages 1581.60 ranges
    5000 requests/second with     2500 ranges averages 2163.70 ranges
   10000 requests/second with     2500 ranges averages 2456.30 ranges
   25000 requests/second with     2500 ranges averages 2499.90 ranges
   50000 requests/second with     2500 ranges averages 2500.00 ranges
    1000 requests/second with     5000 ranges averages 906.20 ranges
    2500 requests/second with     5000 ranges averages 1975.80 ranges
    5000 requests/second with     5000 ranges averages 3169.10 ranges
   10000 requests/second with     5000 ranges averages 4320.90 ranges
   25000 requests/second with     5000 ranges averages 4968.80 ranges
   50000 requests/second with     5000 ranges averages 4999.60 ranges
    1000 requests/second with    10000 ranges averages 950.20 ranges
    2500 requests/second with    10000 ranges averages 2206.70 ranges
    5000 requests/second with    10000 ranges averages 3929.50 ranges
   10000 requests/second with    10000 ranges averages 6326.80 ranges
   25000 requests/second with    10000 ranges averages 9182.60 ranges
   50000 requests/second with    10000 ranges averages 9933.40 ranges
    1000 requests/second with    25000 ranges averages 980.00 ranges
    2500 requests/second with    25000 ranges averages 2381.60 ranges
    5000 requests/second with    25000 ranges averages 4538.20 ranges
   10000 requests/second with    25000 ranges averages 8231.30 ranges
   25000 requests/second with    25000 ranges averages 15809.40 ranges
   50000 requests/second with    25000 ranges averages 21612.70 ranges
    1000 requests/second with    50000 ranges averages 989.70 ranges
    2500 requests/second with    50000 ranges averages 2441.10 ranges
    5000 requests/second with    50000 ranges averages 4755.60 ranges
   10000 requests/second with    50000 ranges averages 9054.00 ranges
   25000 requests/second with    50000 ranges averages 19687.30 ranges
   50000 requests/second with    50000 ranges averages 31579.70 ranges
    1000 requests/second with   100000 ranges averages 995.50 ranges
    2500 requests/second with   100000 ranges averages 2466.80 ranges
    5000 requests/second with   100000 ranges averages 4876.90 ranges
   10000 requests/second with   100000 ranges averages 9514.90 ranges
   25000 requests/second with   100000 ranges averages 22130.30 ranges
   50000 requests/second with   100000 ranges averages 39357.50 ranges
    1000 requests/second with   250000 ranges averages 998.50 ranges
    2500 requests/second with   250000 ranges averages 2487.40 ranges
    5000 requests/second with   250000 ranges averages 4949.70 ranges
   10000 requests/second with   250000 ranges averages 9802.20 ranges
   25000 requests/second with   250000 ranges averages 23799.90 ranges
   50000 requests/second with   250000 ranges averages 45343.90 ranges
    1000 requests/second with   500000 ranges averages 999.20 ranges
    2500 requests/second with   500000 ranges averages 2493.20 ranges
    5000 requests/second with   500000 ranges averages 4973.40 ranges
   10000 requests/second with   500000 ranges averages 9894.80 ranges
   25000 requests/second with   500000 ranges averages 24387.50 ranges
   50000 requests/second with   500000 ranges averages 47575.90 ranges
    1000 requests/second with  1000000 ranges averages 999.60 ranges
    2500 requests/second with  1000000 ranges averages 2496.60 ranges
    5000 requests/second with  1000000 ranges averages 4987.40 ranges
   10000 requests/second with  1000000 ranges averages 9951.80 ranges
   25000 requests/second with  1000000 ranges averages 24687.40 ranges
   50000 requests/second with  1000000 ranges averages 48787.50 ranges
```

spencerkimball commented on storage/replica.go:
```diff
@@ -1638,6 +1645,25 @@ func defaultProposeRaftCommandLocked(r *Replica, p *pendingCmd) error {
+func (r *Replica) setQuiescentLocked(v bool) {
+	if r.mu.quiescent != v {
+		r.mu.quiescent = v
+		if log.V(3) {
```
Is it worthwhile to leave this in?

spencerkimball commented:
LGTM

This is great. Thank you for jumping on this and getting the experiment working.

You're going to need a TODO in here which explores the case of a node going down. Every raft group with a replica on that node will be heartbeating, which could compromise the performance wins of this change, especially for relatively small clusters. We'll need a mechanism to exclude replicas from these checks for quiescense if they reside on nodes which are known to be unresponsive. Also a similar mechanism to wake up all such raft groups when an unresponsive node returns to liveness.

petermattis commented:
Ok, I think this is good to go.

arjunravinarayan commented:
LGTM, I like this. Any bugs are likely to be liveness bugs and not correctness bugs. I think this design fits well with Grouped Heartbeats (queueing up heartbeats and sending them in a single RPC), so we could have both of them together. This might require an RFC to tease out all the edge cases down the road (including the TODO that Spencer asked for above). Ben's point about the lease expiration leads me to think that perhaps the range lease durations should be set at some multiple of the mean latency in the cluster.

---

Review status: 5 of 7 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-:-KRYsLVIP8B7zXZyix2V:b-ntiuvu)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Great work <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>!

---

Reviewed 5 of 7 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 2075 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRYzVN-2mg7aL5etvat:-KRYzVN-2mg7aL5etvau:b-bhrepw) ([raw file](https://github.com/cockroachdb/cockroach/blob/08b555a703d7ae1542d6132461e8159489cfa961/storage/replica.go#L2075)):*
> ```Go
> 			Commit: status.Commit,
> 		}
> 		r.sendRaftMessageRequest(&RaftMessageRequest{
> ```

`sendRaftMessageRequest` has a path that calls `r.withRaftGroup` which looks like a deadlock.

---

*[storage/replica.go, line 2214 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRYzovHbmdLO1Q3wJSM:-KRYzovHbmdLO1Q3wJSN:b-s2ouun) ([raw file](https://github.com/cockroachdb/cockroach/blob/08b555a703d7ae1542d6132461e8159489cfa961/storage/replica.go#L2214)):*
> ```Go
> 		r.mu.Unlock()
> 
> 		if err := r.withRaftGroup(func(raftGroup *raft.RawNode) error {
> ```

This is the callsite I mention above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-:-KRZ-7TtwSRcfM_-N72R:bgsrtlq)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Yes, it will be interesting to see if the decrease in heartbeats corresponds to an increase in QPS. 

I've also added a `COCKROACH_ENABLE_QUIESCENCE` env var (default true) to allow experimentation with turning quiescence on and off.

@spencerkimball Let me think about those cases. I've added a TODO.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1998 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRXg1zjtxy0RNGHO0V2:-KRYgQXQDmg6nOBPyL7p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcacfeff65134dd2f0e274871f069683083239e0/storage/replica.go#L1998)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> if `r.mu.replicaID` is not in `status.Progress`, we shouldn't quiesce. (I think that's possible in weird edge cases during replica changes)
</details>
Done.

---

*[storage/replica.go, line 2001 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRXeVB-Hp_lPN31eAL-:-KRYgR3YiKNc8JG0VJCG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcacfeff65134dd2f0e274871f069683083239e0/storage/replica.go#L2001)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The replica could also be *ahead*: when a new command is proposed, the leader's Progress.Match is advanced to the LastIndex, so this is how we would tell that there are uncommitted entries in the log (especially the dummy entry that concludes an election; others should appear in `pendingCmds`). It's probably worth a comment that we want all of Applied, Commit, LastIndex, and all Progress.Match indexes to be equal, but we get the LastIndex check implicitly via Match (or we could add an explicit check for LastIndex == Commit).
</details>
Done.

---

*[storage/replica.go, line 1657 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRWInA6sg-AfC5HSf28-r2-1657:-KRYgJByfm1lgAerc5En:b830nbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5a0dd1a549c622a5d667c734173e72e318eca1e/storage/replica.go#L1657)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Is it worthwhile to leave this in?
</details>
Leave in the `log.V(3)`? Yes, it was/is useful for debugging.

---

*[storage/replica.go, line 2075 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRYzVN-2mg7aL5etvat:-KRZCnkCYiK3dBnGMldI:b-g47pcp) ([raw file](https://github.com/cockroachdb/cockroach/blob/08b555a703d7ae1542d6132461e8159489cfa961/storage/replica.go#L2075)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `sendRaftMessageRequest` has a path that calls `r.withRaftGroup` which looks like a deadlock.
</details>
It sure does. Refactored so the `ReportUnreachable` call is performed by the caller. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-:-KRYgUprAIocfMcfmaa7:btfvlgu)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Pub-before-push?

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-:-KRZEAO0F1a7Nr6fVzhJ:bfr2f7w)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Yes. Pushed now.

---

Review status: 5 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-:-KRZEcwT8cDF-QU3O9jI:bdk8sa3)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 2233 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRZG82iyUuPmbmHVfJo:-KRZG82iyUuPmbmHVfJp:b-jwnupl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ef66f2da9594663d5b9f4fe5166ca919c21d38/storage/replica.go#L2233)):*
> ```Go
> }
> 
> func (r *Replica) sendRaftMessageRequest(req *RaftMessageRequest) bool {
> ```

Add a quick comment about the bool, please (and perhaps that callers usually want to ReportUnreachable).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 2233 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-KRZG82iyUuPmbmHVfJo:-KRZp8H9rpAs009UFxmC:b-ddqoof) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8ef66f2da9594663d5b9f4fe5166ca919c21d38/storage/replica.go#L2233)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Add a quick comment about the bool, please (and perhaps that callers usually want to ReportUnreachable).
</details>
Done.

Also realized I was still grabbing `r.mu` below and fixed that as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9315#-:-KRZqPyNiXc-W8QwTpZ8:bnfp4nl)*
<!-- Sent from Reviewable.io -->
